crates/accelerate/src/target_transpiler/mod.rs

```diff
@@ -0,0 +1,1253 @@
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+#![allow(clippy::too_many_arguments)]
+
+mod errors;
+mod instruction_properties;
+mod nullable_index_map;
+
+use std::ops::Index;
+
+use ahash::RandomState;
+
+use ahash::HashSet;
+use indexmap::{IndexMap, IndexSet};
+use itertools::Itertools;
+use nullable_index_map::NullableIndexMap;
+use pyo3::{
+    exceptions::{PyAttributeError, PyIndexError, PyKeyError, PyValueError},
+    prelude::*,
+    pyclass,
+    types::{PyDict, PyList, PySet, PyTuple},
+};
+
+use qiskit_circuit::circuit_instruction::convert_py_to_operation_type;
+use qiskit_circuit::operations::{Operation, OperationType, Param};
+use smallvec::SmallVec;
+
+use crate::nlayout::PhysicalQubit;
+
+use errors::TargetKeyError;
+use instruction_properties::InstructionProperties;
+
+use self::exceptions::TranspilerError;
+
+mod exceptions {
+    use pyo3::import_exception_bound;
+    import_exception_bound! {qiskit.exceptions, QiskitError}
+    import_exception_bound! {qiskit.transpiler.exceptions, TranspilerError}
+}
+
+// Custom types
+type Qargs = SmallVec<[PhysicalQubit; 2]>;
+type GateMap = IndexMap<String, PropsMap, RandomState>;
+type PropsMap = NullableIndexMap<Qargs, Option<InstructionProperties>>;
+type GateMapState = Vec<(String, Vec<(Option<Qargs>, Option<InstructionProperties>)>)>;
+
+#[derive(Debug, Clone, FromPyObject)]
+pub(crate) enum TargetOperation {
+    Normal(NormalOperation),
+    Variadic(PyObject),
+}
+
+impl IntoPy<PyObject> for TargetOperation {
+    fn into_py(self, py: Python<'_>) -> PyObject {
+        match self {
+            Self::Normal(normal) => normal.into_py(py),
+            Self::Variadic(variable) => variable,
+        }
+    }
+}
+
+impl ToPyObject for TargetOperation {
+    fn to_object(&self, py: Python<'_>) -> PyObject {
+        match self {
+            Self::Normal(normal) => normal.to_object(py),
+            Self::Variadic(variable) => variable.clone_ref(py),
+        }
+    }
+}
+
+impl TargetOperation {
+    fn num_qubits(&self) -> u32 {
+        match &self {
+            Self::Normal(normal) => normal.operation.num_qubits(),
+            Self::Variadic(_) => {
+                unreachable!("'num_qubits' property is reserved for normal operations only.")
+            }
+        }
+    }
+
+    fn params(&self) -> &[Param] {
+        match &self {
+            TargetOperation::Normal(normal) => normal.params.as_slice(),
+            TargetOperation::Variadic(_) => &[],
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct NormalOperation {
+    pub operation: OperationType,
+    pub params: SmallVec<[Param; 3]>,
+    op_object: PyObject,
```

Comment:
Yes, that is exactly the purpose of this. So that we don't have to reconstruct the `Operation` object. However, I understand that Jake's changes in #12730 might reshape the way this is done. So storing the `op_object` might not be necessary after that merges, so maybe we should wait for that PR.